design/design-doc-integration-tests.md

```diff
@@ -0,0 +1,333 @@
+
+The features above are already handled by robust platforms that are currently
+serving their purpose effectively. Addressing them in the initial implementation
+of this framework would require more effort than the return they would provide.
+
+## Definition of a Successful Test
+
+As mentioned above, the framework should unopinionated about the desired outcome
+of a test. Every test should be able to define what it deems to be successful
+and the framework should merely verify that that the successful conditions are
+met or not.
+
+### Definition of a Testable Unit
+
+Any controller that that is able to successfully register with the Kubernetes
+control plane should be testable. Any valid CustomResourceDefinitions that are
+able to be registered with the [Scheme] of the Kubernetes API should be
+testable.
+
+## Proposal
+
+The following sections describe a testing framework that utilizes the Go
+[testing] package to execute tests against a local Kubernetes control plane. The
+framework is designed such that it can be minimally implemented in the
+short-term, deferring most of the execution and configuration to the tests that
+are written for each implementation. It is intended to evolve over time to
+reduce the amount of configuration required by each implementation, while still
+allowing flexibility to test any scenario.
+
+### System Design
+
+The framework should be made up of two core components:
+
+- `Job`: a testing environment. Every job runs the full setup and tear-down of a
+  local Kubernetes control plane. A job is made up of configuration, which
+  informs the framework of how to setup the control plane, and `Tests` which
+  test actual execution in the control plane.
+- `Test`: a test is an action or set of actions to execute against the
+  Kubernetes control plane. It could involve one or more steps, each of which
+  may cause the test to fail. Each test should be a logical operation, meaning
+  that tests that involve multiple steps should still be checking a single
+  logical condition (i.e. a `TestCreateSubnetSuccessful` may create the
+  `Subnet`, then wait to make sure that its `ConditionedStatus` is `Ready:
+  True`). 
+
+#### Jobs
+
+A `Job` contains the following components. `*Config` will likely evolve over
+time, but will start with just defining the `SyncPeriod` for the controller
+`manager`.
+
+```go
+// A Job is a set of tests that run in a custom configured local Kubernetes environment.
+type Job struct {
+    name             string
+    description      string
+    cfg              *Config
+    runner           *testing.T
+    setupWithManager setupWithManagerFunc
+    addToScheme      addToSchemeFunc
+    jobClean         jobCleanFunc
+    tests            []Test
+}
+```
+
+A job's primary method is `func (j *Job) Run() error` which executes the
+following broadly defined steps:
+
+1. Starting the API Server
+
+The Kubernetes [controller-runtime] project provides an [envtest] package for
```

Comment:
@negz the more I work with this, the more I don't really see a lot of value in wrapping the `envtest` package. I think we will hardly ever be loading the binaries from env variables directly and will likely be using a remote cluster or KIND cluster. I think removing this dependency and just injecting a `client.Config` into the framework at runtime will be better option, and we will have more control over what the API looks like as this framework evolves. Does that sound like the right direction to you?

Reply:
One area I already have been able to improve since removing `envtest` is the support for `v1` CRDs. We are still currently using only `v1beta1` CRDs across the Crossplane ecosystem, but the ability to enhance the framework at our own desired paces is a big win IMO.

Reply:
Circling back on this, I have somewhat changed my tune in that all of the controllers are using `controller-runtime` so it makes sense to use it here as well. I have made significant updates to the API though and am feeling better about the architecture: https://github.com/hasheddan/athodyd